### PR TITLE
Check client instrumentation

### PR DESCRIPTION
**Pull request by pakrym**
Adds a guard rail that verifies that at least one client was instrumented if the test has any recordings.

### CODE REVIEW

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 1]
```diff
@@ -143,16 +145,30 @@ public virtual void StartTestRecording()
...
 
         [TearDown]
         public virtual void StopTestRecording()
         {
-            bool save = TestContext.CurrentContext.Result.Outcome.Status == TestStatus.Passed;
+            bool testPassed = TestContext.CurrentContext.Result.Outcome.Status == TestStatus.Passed;
+
+            if (ValidateClientInstrumentation && testPassed)
+            {
+                throw new InvalidOperationException("The test didn't instrument any clients but had recordings. Please use call InstrumentClient for the client being recorded.");
```
Comment (christothes) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:158]:
```suggestion
                throw new InvalidOperationException("The test didn't instrument any clients but had recordings. Please call InstrumentClient for the client being recorded.");
```

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 2]
```diff
@@ -143,16 +145,30 @@ public virtual void StartTestRecording()
...
+            ValidateClientInstrumentation = Recording.HasRequests;
         }
 
         [TearDown]
         public virtual void StopTestRecording()
         {
-            bool save = TestContext.CurrentContext.Result.Outcome.Status == TestStatus.Passed;
+            bool testPassed = TestContext.CurrentContext.Result.Outcome.Status == TestStatus.Passed;
+
+            if (ValidateClientInstrumentation && testPassed)
```
Comment (christothes) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:156]: I think this logic might be more clear if `ValidateClientInstrumentation` was changed to `IsClientInstrumented` and this line changes to:
```suggestion
            if (Recording.HasRequests && IsClientInstrumented && testPassed)
```

Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:156]: Looking further down, I see we use this to set whether instrumentation should be validated. Perhaps just differentiating between "ShouldValidate" and "IsInstrumented" would be better then?

**Review by christothes — APPROVED** [review]

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs (hunk 2 above)
Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:156]: You can't check `Recording.HasRequests` here as requests are removed from the recording during the playback. So `Recording.HasRequests` would always be false. That's why the logic looks a bit strange.

**Pull request merged**